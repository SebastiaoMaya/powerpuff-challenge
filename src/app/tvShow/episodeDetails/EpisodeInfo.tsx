import { IconButton } from "@chakra-ui/button";
import { ExternalLinkIcon } from "@chakra-ui/icons";
import { Stack, Text } from "@chakra-ui/layout";
import { Image } from "app/shared";
import i18next from "i18next";
import React from "react";
import { Episode } from "../models";

export const EpisodeInfo: React.FC<Episode> = ({
  name,
  url,
  summary,
  image,
}) => (
  <Stack flexDir="column">
    <Stack flexDir="row" alignItems="center">
      <Text fontSize={["lg", "7xl"]} pr={2}>
        {name}
      </Text>
      {url && (
        <IconButton
          colorScheme="blue"
          aria-label="Open episode url"
          icon={<ExternalLinkIcon />}
          onClick={() => window.open(url, "_blank")}
        />
      )}
    </Stack>
    <Text
      dangerouslySetInnerHTML={{
        __html:
          summary && summary.length !== 0
            ? summary
            : i18next.t("tvshow:episodeInfo.noSummary"),
      }}
      fontSize={["md", "lg"]}
    />
    <Image src={image?.medium} w={"50%"} />
  </Stack>
);
