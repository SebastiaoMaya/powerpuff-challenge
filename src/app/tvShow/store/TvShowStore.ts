import { action, computed, flow, makeObservable, observable } from "mobx";
import { TvShow, Episode } from "../models";
import { fetchTvShowService, fetchEpisodesListService } from "../services";

export class TvShowStore {
  @observable private _tvShow: TvShow | undefined = undefined;
  @observable private _episodesList: Episode[] = [];

  @action.bound setTvShow(tvShow: TvShow) {
    this._tvShow = tvShow;
  }

  fetchTvShow = flow(function* (this: TvShowStore) {
    yield fetchTvShowService().then((r) => {
      this.setTvShow(r);
      this.fetchEpisodesList(r.id);
    });
  }).bind(this);

  fetchEpisodesList = flow(function* (this: TvShowStore, tvShowId: string) {
    this._episodesList = yield fetchEpisodesListService(tvShowId);
  }).bind(this);

  @computed get tvShow(): TvShow | undefined {
    return this._tvShow;
  }

  @computed get episodesList(): Episode[] {
    return this._episodesList;
  }

  constructor() {
    makeObservable(this);
  }
}
