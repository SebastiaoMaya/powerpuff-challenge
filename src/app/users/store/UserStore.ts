import { computed, flow, makeObservable, observable } from "mobx";
import { User } from "../models";
import { doLogin } from "../services";

export class UserStore {
  @observable private _user: User | undefined = undefined;

  login = flow(function* (this: UserStore) {
    this._user = yield doLogin();
  }).bind(this);

  logout = flow(function* (this: UserStore) {
    this._user = yield undefined;
  }).bind(this);

  @computed get user(): User | undefined {
    return this._user;
  }

  constructor() {
    makeObservable(this);
  }
}
