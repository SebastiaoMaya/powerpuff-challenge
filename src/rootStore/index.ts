export { useRootStoreContext } from "./context";
export { RootStoreProvider } from "./provider";
