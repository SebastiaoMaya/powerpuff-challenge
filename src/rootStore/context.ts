import { UserStore } from "app/users/store/UserStore";
import React, { useContext } from "react";

interface StoreContextProps {
  userStore: UserStore;
}

export const RootStoreContext = React.createContext<StoreContextProps>(
  {} as StoreContextProps
);

export const useRootStoreContext = () => useContext(RootStoreContext);
