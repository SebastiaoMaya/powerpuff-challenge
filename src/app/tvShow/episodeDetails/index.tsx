import {
  Breadcrumb,
  BreadcrumbItem,
  BreadcrumbLink,
} from "@chakra-ui/breadcrumb";
import { ChevronRightIcon } from "@chakra-ui/icons";
import { Text } from "@chakra-ui/layout";
import i18next from "i18next";
import { toJS } from "mobx";
import { observer } from "mobx-react";
import React, { useState, useEffect, useCallback } from "react";
import { Link } from "react-router-dom";
import { useRootStoreContext } from "rootStore";
import { Episode } from "../models";
import { EpisodeInfo } from "./EpisodeInfo";

interface Props {
  id: string;
}

const EpisodeDetailsBase: React.FC<Props> = ({ id }) => {
  const { tvShowStore } = useRootStoreContext();
  const [episode, setEpisode] = useState<Episode | undefined>();

  const getEpisodeById = useCallback(
    (id: string) => {
      setEpisode(
        tvShowStore.episodesList.find((e) => toJS(e).id.toString() === id)
      );
    },
    [tvShowStore.episodesList]
  );

  useEffect(() => {
    !tvShowStore.tvShow && tvShowStore.fetchTvShow();
    getEpisodeById(id);
  }, [tvShowStore, id, getEpisodeById]);

  return (
    <>
      <Breadcrumb
        spacing="8px"
        separator={<ChevronRightIcon color="gray.500" />}>
        <BreadcrumbItem>
          <BreadcrumbLink as={Link} to="/">
            Home
          </BreadcrumbLink>
        </BreadcrumbItem>
        <BreadcrumbItem isCurrentPage>
          <BreadcrumbLink>
            {episode?.name ??
              i18next.t("tvshow:episodeDetails:episodeNotFound")}
          </BreadcrumbLink>
        </BreadcrumbItem>
      </Breadcrumb>
      {episode ? (
        <EpisodeInfo {...episode} />
      ) : (
        <Text fontSize={["lg", "7xl"]}>
          {i18next.t("tvshow:episodeDetails:episodeNotFound")}
        </Text>
      )}
    </>
  );
};

export const EpisodeDetails = observer(EpisodeDetailsBase);
