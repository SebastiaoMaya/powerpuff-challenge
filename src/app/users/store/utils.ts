import { useRootStoreContext } from "rootStore";

export const useIsUserLoggedIn = () => {
  const { userStore } = useRootStoreContext();

  return !!userStore.user;
};
