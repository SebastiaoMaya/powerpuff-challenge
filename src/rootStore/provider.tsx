import { TvShowStore } from "app/tvShow/store";
import React from "react";
import { RootStoreContext } from "./context";

interface Props {
  children: React.ReactNode;
}

export const RootStoreProvider: React.FC<Props> = ({ children }) => {
  const tvShowStore = new TvShowStore();

  return (
    <RootStoreContext.Provider value={{ tvShowStore }}>
      {children}
    </RootStoreContext.Provider>
  );
};
