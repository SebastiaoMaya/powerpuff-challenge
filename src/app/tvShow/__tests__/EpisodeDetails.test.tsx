import React from "react";
import { render } from "@testing-library/react";
import MockAdapter from "axios-mock-adapter";
import axios from "network";
import { EpisodeDetails } from "../episodeDetails";
import {
  MOCK_EPISODES_LIST_RESPONSE,
  MOCK_TV_SHOW_RESPONSE,
} from "../../../__mocks__/mocks";
import { RootStoreContext } from "rootStore/context";
import { TvShowStore } from "../store";
import { BrowserRouter } from "react-router-dom";

const renderEpisodeDetails = () =>
  render(
    <RootStoreContext.Provider value={{ tvShowStore: new TvShowStore() }}>
      <BrowserRouter>
        <EpisodeDetails id={"657308"} />
      </BrowserRouter>
    </RootStoreContext.Provider>
  );

describe("EpisodeDetails", () => {
  let mockAxios: MockAdapter;

  beforeEach(() => {
    // mockAxios = new MockAdapter(axios);
    // mockAxios
    //   .onGet("https://api.tvmaze.com/singlesearch/shows?q=powerpuff")
    //   .reply(200, MOCK_TV_SHOW_RESPONSE);
    // mockAxios
    //   .onGet("https://api.tvmaze.com/shows/6771/episodes")
    //   .reply(200, MOCK_EPISODES_LIST_RESPONSE);
  });

  afterEach(() => {
    //mockAxios.reset();
  });

  it("renders EpisodeDetails", () => {
    const { container } = renderEpisodeDetails();

    expect(container).toMatchSnapshot();
  });

  //   it('renders titles', () => {
  //     const { getByText } = renderConsents(legalScenarioExplicitText);
  //     expect(getByText(legalScenarioExplicitText.documentList[0].document.documentText1)).toBeTruthy();
  //   });

  //   it('is required when acceptanceRequired=true', () => {
  //     const { container } = renderConsents(legalScenarioExplicitText);
  //     const requiredItems = container.querySelectorAll('.gl-form-item--required');
  //     const requiredText = getNodesText(requiredItems);
  //     expect(requiredText).toMatchObject([legalScenarioExplicitText.documentList[1].document.documentText1]);
  //   });

  //   it('is pre-checked when acceptancePreSelected=true', () => {
  //     const onChange = jest.fn();
  //     renderConsents(legalScenarioExplicitPreChecked, onChange);

  //     expect(onChange).toHaveBeenCalledTimes(1);
  //     expect(onChange.mock.calls[0][0].acceptances[0].accepted).toBe(true);
  //   });

  //   it('shows errors for mandatory checkboxes', () => {
  //     const { getByText } = renderConsentsTouched(legalScenarioExplicitText);

  //     const errorText = getByText(errorrequired);
  //     expect(errorText).toBeTruthy();
  //   });
});
