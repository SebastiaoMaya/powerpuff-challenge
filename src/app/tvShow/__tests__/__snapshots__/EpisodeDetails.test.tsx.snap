// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EpisodeDetails renders EpisodeDetails 1`] = `
<div>
  <nav
    aria-label="breadcrumb"
    class="chakra-breadcrumb css-0"
  >
    <ol
      class="chakra-breadcrumb__list css-0"
    >
      <li
        class="chakra-breadcrumb__list-item css-18biwo"
      >
        <a
          class="chakra-breadcrumb__link css-0"
          href="/"
        >
          Home
        </a>
        <span
          class="css-syhbcg"
          role="presentation"
        >
          <svg
            class="chakra-icon css-2u1jce"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
              fill="currentColor"
            />
          </svg>
        </span>
      </li>
      <li
        class="chakra-breadcrumb__list-item css-18biwo"
      >
        <span
          aria-current="page"
          class="chakra-breadcrumb__link css-0"
        />
      </li>
    </ol>
  </nav>
  <p
    class="chakra-text css-1iec28v"
  />
</div>
`;
